File: leetcode/128-longest-consecutive-sequence.cpp
/*
Given an unsorted array of integers, find the length of the longest consecutive elements sequence.

For example,
Given [100, 4, 200, 1, 3, 2],
The longest consecutive elements sequence is [1, 2, 3, 4]. Return its length: 4.

Your algorithm should run in O(n) complexity.
*/

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_map<int, int> heads;
        unordered_map<int, int> tails;
        for (auto num : nums) {
            auto i = heads.find(num);
            if (i == heads.end()) {
                heads[num] = 1;
                tails[num] = num;
                auto next_head = heads.find(num + 1);
                if (next_head != heads.end()) {
                    tails[num + next_head->second] = num;
                    heads[num] = 1 + next_head->second;
                }
                auto prev_tail = tails.find(num - 1);
                if (prev_tail != tails.end()) {
                    auto prev_head = prev_tail->second;
                    tails[num + heads[num] - 1] = prev_head;
                    heads[prev_head] += heads[num];
                }
            }
        }
        int counter = 0;
        for (auto i : heads) {
            counter = max(counter, i.second);
        }
        return counter;
    }
};

File: leetcode/128-longest-consecutive-sequence.rb
require 'set'

# @param {Integer[]} nums
# @return {Integer}
def longest_consecutive(nums)
  heads = Hash.new # head -> tail
  tails = Hash.new # tail -> head
  all = Set.new
  nums.each do |num|
    next if all.include? num
    all.add num
    old_tail = heads[num + 1]
    old_head = tails[num - 1]
    if (not old_tail.nil?) and (not old_head.nil?)
      heads[old_head] = old_tail
      tails[old_tail] = old_head
      heads.delete num + 1
      tails.delete num - 1
    elsif not old_head.nil?
      heads[old_head] = num
      tails[num] = old_head
      tails.delete num - 1
    elsif not old_tail.nil?
      tails[old_tail] = num
      heads[num] = old_tail
      heads.delete num + 1
    elsif (not heads.include? num) and (not tails.include? num)
      heads[num] = num
      tails[num] = num
    end
  end
  heads.inject(0) { |max, x| [max, x[1] - x[0] + 1].max }
end

